pkg/clustermesh/common/namespacewatcher/watcher.go

```diff
@@ -0,0 +1,270 @@
+			result.Insert(ns.Name)
+		}
+	}
+
+	return result
+}
+
+func (nw *namespaceWatcher) IsGlobalService(obj interface {
+	GetAnnotations() map[string]string
+	GetNamespace() string
+}) bool {
+	// First check if service has the global annotation
+	if !annotation.GetAnnotationIncludeExternal(obj) {
+		return false
+	}
+
+	// If namespace filtering is not active, service is global.
+	// If namespace filtering is active, also check if service is in a global namespace
+	return nw.IsGlobalNamespace(obj.GetNamespace())
+}
+
+func (nw *namespaceWatcher) IsSharedGlobalService(obj interface {
+	GetAnnotations() map[string]string
+	GetNamespace() string
+}) bool {
+	// Service must be global first
+	if !nw.IsGlobalService(obj) {
+		return false
+	}
+
+	// Then check if it's also shared
+	return annotation.GetAnnotationShared(obj)
+}
+
+func (nw *namespaceWatcher) processNamespaceChange(processors []NamespaceProcessor, namespace string, isGlobal bool) {
+
+	// Normal operation: process asynchronously
+	for _, processor := range processors {
+		processor.OnNamespaceGlobalChange(namespace)
+	}
+}
+
+func (nw *namespaceWatcher) updateNamespace(ns *slim_corev1.Namespace) {
+	nw.mu.Lock()
+
+	// Capture the previous state by querying the resource store
+	wasGlobal := nw.isNamespaceGlobalByAnnotation(ns.Name)
+
+	var processors = make([]NamespaceProcessor, len(nw.processors))
+	copy(processors, nw.processors)
+
+	nw.mu.Unlock()
+
+	// Calculate current global status after the namespace update
+	// We need to check the current filtering state and annotation status
+	isGlobal := nw.calculateGlobalStatus(ns)
+
+	// Check if this specific namespace's status changed
+	needsProcessing := wasGlobal != isGlobal
+
+	if needsProcessing {
+		nw.logger.Info("Namespace global status changed",
+			logfields.K8sNamespace, ns.Name,
+			logfields.IsGlobal, isGlobal,
+		)
+
+		// Process the namespace change directly without acquiring additional locks
+		nw.processNamespaceChange(processors, ns.Name, isGlobal)
+	}
+}
+
+// hasOtherAnnotatedNamespaces checks if there are any other namespaces (besides the given one) that have annotations
+func (nw *namespaceWatcher) hasOtherAnnotatedNamespaces(excludeNamespace string) bool {
+	allNamespaces := nw.nsStore.List()
+	for _, ns := range allNamespaces {
+		if ns.Name != excludeNamespace {
+			if _, hasAnnotation := annotation.Get(ns, annotation.GlobalNamespace); hasAnnotation {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// calculateGlobalStatus determines if a namespace should be global based on its annotation and filtering state.
+// Returns true if the namespace is global.
+func (nw *namespaceWatcher) calculateGlobalStatus(ns *slim_corev1.Namespace) bool {
+	// Get global annotation for this namespace.
+	annotationValue, hasAnnotation := annotation.Get(ns, annotation.GlobalNamespace)
+
+	return nw.config.DefaultGlobalNamespace || (hasAnnotation && strings.ToLower(annotationValue) == "true")
+}
```

Comment:
The logic here not correct either, and also diverged with respect to `IsGlobalNamespace`, which should be refactored to make use of this method (once fixed).